internal/pkg/provider/provision.go

```diff
@@ -0,0 +1,276 @@
+			if err != nil {
+				return err
+			}
+
+			vm.Spec.Running = pointer.To(true)
+
+			if vm.Spec.Template == nil {
+				vm.Spec.Template = &kvv1.VirtualMachineInstanceTemplateSpec{
+					Spec: kvv1.VirtualMachineInstanceSpec{
+						Domain: kvv1.DomainSpec{
+							Resources: kvv1.ResourceRequirements{
+								Requests: v1.ResourceList{},
+							},
+						},
+					},
+				}
+			}
+
+			vm.Spec.Template.Spec.Domain.Firmware = &kvv1.Firmware{
+				UUID: types.UID(pctx.State.TypedSpec().Value.Uuid),
+			}
+
+			vm.Spec.Template.Spec.Architecture = data.Architecture
+			vm.Spec.Template.Spec.Domain.CPU = &kvv1.CPU{
+				Cores: uint32(data.Cores),
+			}
+
+			vm.Spec.Template.Spec.Domain.Resources.Requests[v1.ResourceMemory] = *resource.NewQuantity(int64(data.Memory)*1024*1024, resource.DecimalSI)
+
+			vm.Spec.Template.Spec.Domain.Devices = kvv1.Devices{
+				Disks: []kvv1.Disk{
+					{
+						Name:      "kv",
+						BootOrder: pointer.To(uint(1)),
+						DiskDevice: kvv1.DiskDevice{
+							Disk: &kvv1.DiskTarget{
+								Bus: kvv1.DiskBusVirtio,
+							},
+						},
+					},
+				},
+			}
+
+			vm.Spec.Template.Spec.Volumes = []kvv1.Volume{
+				{
+					Name: "kv",
+					VolumeSource: kvv1.VolumeSource{
+						DataVolume: &kvv1.DataVolumeSource{
+							Name: pctx.GetRequestID(),
+						},
+					},
+				},
+			}
+
+			vm.Spec.DataVolumeTemplates = []kvv1.DataVolumeTemplateSpec{
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: pctx.GetRequestID(),
+					},
+					Spec: cdiv1.DataVolumeSpec{
+						PVC: &v1.PersistentVolumeClaimSpec{
+							VolumeMode: pointer.To(v1.PersistentVolumeBlock),
+							AccessModes: []v1.PersistentVolumeAccessMode{
+								v1.ReadWriteOnce,
+							},
+							Resources: v1.VolumeResourceRequirements{
+								Requests: v1.ResourceList{
+									v1.ResourceStorage: resource.MustParse(fmt.Sprintf("%dGi", data.DiskSize)),
+								},
+							},
+						},
+						Source: &cdiv1.DataVolumeSource{
+							PVC: &cdiv1.DataVolumeSourcePVC{
+								Name:      pctx.State.TypedSpec().Value.VolumeId,
+								Namespace: p.namespace,
+							},
+						},
+					},
+				},
+			}
+
+			if vm.Name == "" {
+				vm.Name = pctx.GetRequestID()
+				vm.Namespace = p.namespace
+
+				if err = p.k8sClient.Create(ctx, vm); err != nil {
+					return err
+				}
+			} else {
+				if err = p.k8sClient.Update(ctx, vm); err != nil && !errors.IsConflict(err) {
+					return err
+				}
+			}
+
+			return provision.NewRetryInterval(time.Second * 10)
+		}),
+	}
+}
+
+// Deprovision implements infra.Provisioner.
+func (p *Provisioner) Deprovision(ctx context.Context, logger *zap.Logger, _ *resources.Machine, machineRequest *infra.MachineRequest) error {
+	err := p.k8sClient.Delete(ctx, &kvv1.VirtualMachine{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      machineRequest.Metadata().ID(),
+			Namespace: p.namespace,
+		},
+	})
+	if err != nil && !errors.IsNotFound(err) {
+		return err
+	}
+
+	logger.Info("machine deprovisioned")
```

Comment:
should we wait for the machine to be actually removed?

Reply:
We're removing VM from the KubeVIrt, so I guess it blocks the execution until the resource is gone.
Then when there's no VM it's deprovisioned from the the provider perspective

Reply:
It doesn't block iirc, it just "marks as deleted", so it might come as a surprise if you delete 5 VMs and can't create 5 new VMs, as you're out of capacity. 

So I guess like Create waits for the resource to be ready, we should wait for the resource to be actually deleted here.